### PR TITLE
Add ability to filter by content of fields and toggle reply trees

### PR DESCRIPTION
**Pull request by IshaniSanturkar**
In this PR, I:
- added the ability to filter comments by the contents of name, email and comment fields
- added a button to toggle reply trees in and out 
Screenshot 2020-06-05 at 4 00 48 PM

### CODE REVIEW

File: portfolio/index.yaml [hunk 1]
```diff
@@ -53,5 +53,173 @@ indexes:
...
+  properties:
+  - name: name
+  - name: time
+    direction: desc
+
+- kind: Comment
+  ancestor: no
+  properties:
+  - name: name
+  - name: name
```
Comment (fenghaolw) [inline portfolio/index.yaml:70]: I don't understand this. Why is this needed?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -62,7 +64,7 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
    * Returns a list containing atmost maxComments top-level queries and all their replies. 
    * The top-level queries are sorted by sortMetric in sortOrder
    */ 
-  private void populateRootComments(ArrayList<UserComment> comments, int maxComments, String sortOrder, String sortMetric) {
+  private void populateRootComments(ArrayList<UserComment> comments, int maxComments, String sortOrder, String sortMetric, String filterMetric, String filterText) {
```
Comment (fenghaolw) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: style guide suggests a 100-char limit
https://google.github.io/styleguide/javaguide.html#s4.4-column-limit

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -182,13 +186,26 @@ function constructReplyTree(comment, commentTree, margin) {
...
     for (const child of children) {
       const subTree = constructReplyTree(child, commentTree, newMargin);
       replyTree.appendChild(subTree);
     }
     replyTree.style.marginLeft = `${margin}px`;
+    const toggleButton = document.createElement("button");
+    toggleButton.classList.add("material-icons", "togglereply");
+    toggleButton.innerText = "unfold_less";
+    toggleButton.onclick = () => {
+        if(replyTree.classList.contains("hiddenreplytree")) {
```
Comment (fenghaolw) [inline portfolio/src/main/webapp/script.js]: nit: indentation

File: portfolio/index.yaml (hunk 1 above)
Comment (IshaniSanturkar) [reply portfolio/index.yaml:70]: It's because for more complex queries, you need to create an index file with all the possible configurations of queries in advance. Since I'm filtering by 3 parameters and sorting by 4 in 2 orders there's a lot of entries :(. I was wondering whether I should add this feature because of this.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -46,12 +46,15 @@
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int maxComments = Integer.parseInt(UtilityFunctions.getFieldFromResponse(request, "maxcomments", defaultMaxComment));
+    int maxComments = Integer.parseInt(UtilityFunctions.getFieldFromResponse(request, "maxcomments"
```
Comment (EatingW) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Sorry you couldn't auto format your Java code :( 

According to https://google.github.io/styleguide/javaguide.html#s4.5.1-line-wrapping-where-to-break, the comma stays attached to the preceding element. 

In your case I think it would look nicer if you line break right after `getFieldFromResponse`, so that the arguments stays together. Something like:
```
int maxComments = Integer.parseInt(UtilityFunctions.getFieldFromResponse(
    request, "maxcomments", defaultMaxComment));
```
You can also play around with other ways for line breaking and see which one you like. 
```
int maxComments = Integer.parseInt(
    UtilityFunctions.getFieldFromResponse(request, "maxcomments", defaultMaxComment));
```
```
int maxComments = Integer.parseInt(
    UtilityFunctions.getFieldFromResponse(
        request, "maxcomments", defaultMaxComment));
```
```
int maxComments = Integer.parseInt(
    UtilityFunctions.getFieldFromResponse(
        request, 
        "maxcomments", 
        defaultMaxComment));
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -62,7 +65,8 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
    * Returns a list containing atmost maxComments top-level queries and all their replies. 
    * The top-level queries are sorted by sortMetric in sortOrder
    */ 
-  private void populateRootComments(ArrayList<UserComment> comments, int maxComments, String sortOrder, String sortMetric) {
+  private void populateRootComments(ArrayList<UserComment> comments, int maxComments
```
Comment (EatingW) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: If a method definition gets too long, the parameters are generally listed line-by-line. i.e.
```
private void populateRootComments(
    ArrayList<UserComment> comments, 
    int maxComments, 
    String sortOrder,
    String sortMetric, 
    String filterMetric, 
    String filterText) {
```

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -182,13 +186,26 @@ function constructReplyTree(comment, commentTree, margin) {
...
     let replyTree = document.createElement("ul");
-    replyTree.className = "comments";
+    replyTree.classList.add("comments", "visiblereplytree");
     const newMargin = margin + 20;
     for (const child of children) {
       const subTree = constructReplyTree(child, commentTree, newMargin);
       replyTree.appendChild(subTree);
     }
     replyTree.style.marginLeft = `${margin}px`;
+    const toggleButton = document.createElement("button");
```
Comment (EatingW) [inline portfolio/src/main/webapp/script.js]: This constructReplyTree function is getting a bit long. Separate the creation of toggleButton into its own function?

File: portfolio/src/main/webapp/comments.html [hunk 7]
```diff
@@ -66,6 +66,14 @@ <h2>Comment History</h2>
       <span class="commentparam">
         <button onclick="clearComments()">Clear</button>
       </span>
+      <div id="filterdiv">
+        <select name="filterby" id="filterby" onchange="loadComments()">
```
Comment (EatingW) [inline portfolio/src/main/webapp/comments.html:71]: The same way the label "Sort by:" preceded the select element, add a "Filter by:" text so that it makes it clear for the users what the select element is for?

**Review by EatingW — CHANGES_REQUESTED** [review]
Looks good! Mostly some formatting nits.

**Review by EatingW — CHANGES_REQUESTED** [review]
nit: the Java formatting in the rest of DataServlet.java, ReplyServlet.java, UserComment.java, UtilityFunctions.java also needs to be fixed, similar to how you fixed `doGet` and `populateRootComments` in DataServlet.java

**Review by EatingW — APPROVED** [review]
LGTM!
